File: task-scheduler-backend/src/graphql/types/notification.rs
use async_graphql::*;
use chrono::{DateTime, Utc};
use serde_json::Value as JsonValue;
use uuid::Uuid;

#[derive(SimpleObject)]
pub struct Notification {
    pub id: ID,
    pub user_id: ID,
    pub project_id: Option<ID>,
    pub sender_id: Option<ID>,
    pub type_: String,
    pub reference_type: String,
    pub reference_id: ID,
    pub message: String,
    pub action: String,
    pub metadata: JsonValue,
    pub is_read: bool,
    pub created_at: DateTime<Utc>,
    pub read_at: Option<DateTime<Utc>>,
}

#[derive(InputObject)]
pub struct CreateNotificationInput {
    pub user_id: ID,
    pub project_id: Option<ID>,
    pub sender_id: Option<ID>,
    pub type_: String,
    pub reference_type: String,
    pub reference_id: ID,
    pub message: String,
    pub action: String,
    pub metadata: Option<JsonValue>,
}

#[derive(SimpleObject)]
pub struct NotificationCount {
    pub total: i64,
    pub unread: i64,
}

impl From<crate::db::models::Notification> for Notification {
    fn from(notif: crate::db::models::Notification) -> Self {
        Self {
            id: notif.notification_id.to_string().into(),
            user_id: notif.user_id.to_string().into(),
            project_id: notif.project_id.map(|id| id.to_string().into()),
            sender_id: notif.sender_id.map(|id| id.to_string().into()),
            type_: notif.type_,
            reference_type: notif.reference_type,
            reference_id: notif.reference_id.to_string().into(),
            message: notif.message,
            action: notif.action,
            metadata: notif.metadata,
            is_read: notif.is_read,
            created_at: notif.created_at,
            read_at: notif.read_at,
        }
    }
} 

File: web/src/types/notification.ts
export type NotificationType = 'TASK_ASSIGNED' | 'TASK_REASSIGNED' | 'TASK_COMPLETED' | 'TASK_OVERDUE' | 'COMMENT_MENTION' | 'TASK_COMMENT' | 'COMMENT_MENTION';
// column_name               |
// --------------------------+
// notification_id           |
// user_id                   |
// type                      |
// reference_type            |
// reference_id              |
// message                   |
// is_read                   |
// created_at                |
// project_id                |
// sender_id                 |
// action                    |
// metadata                  |
export interface BackendNotification {
  notification_id?: string;
  user_id: string;
  message: string;
  type: string;
  reference_type: string;
  reference_id: string;
  is_read: boolean;
  created_at: string;
  project_id?: string;
  sender_id?: string;
  action: string;
  metadata?: {
    project_id: string;
    task_id: string;
    comment_id?: string;
    task_title?: string;
  };
}

// Interface for UI components - camelCase for frontend use
export interface Notification {
  id: string;
  userId: string;
  message: string;
  type: NotificationType;
  isRead: boolean;
  createdAt: string;
  projectId?: string;
  taskId?: string;
  commentId?: string;
  senderId?: string;
  metadata?: {
    project_id: string;
    task_id: string;
    comment_id?: string;
    task_title?: string;
  };
}

export interface NotificationState {
  notifications: Notification[];
  unreadCount: number;
}

export interface NotificationCount {
  total: number;
  unread: number;
}

export interface CreateNotificationInput {
  userId: string;
  message: string;
  type: NotificationType;
  projectId?: string;
  taskId?: string;
  commentId?: string;
  senderId?: string;
}

// Response interfaces updated for new naming convention
export interface NotificationQueryResponse {
  getNotifications?: BackendNotification[];
  notifications?: BackendNotification[];  // Fallback for backward compatibility
}

export interface NotificationCountQueryResponse {
  getNotificationCount?: NotificationCount;
  notificationCount?: NotificationCount;  // Fallback for backward compatibility
}
